Type-class hierarchy for cancelation

alexandru opened the pull request:
This PR is the second proposal of a type class hierarchy that supports cancellation, following the development of the cancelable `cats.effect.IO` in #121

This proposal introduces two new and separate type classes, `Concurrent` and `ConcurrentEffect`, in addition to the existing `Async` and `Effect`.

---

Also see the first proposal, from #126, that modifies `Async` and `Effect`, thus making cancellation optional (as an optimization that the data type may, or may not support).

---

The new hierarchy:

![image](https://user-images.githubusercontent.com/11753/36951388-d7c66b04-200b-11e8-9c92-84ffa9f0e987.png)

The new `Concurrent` type class:

```scala
trait Concurrent[F[_]] extends Async[F] {

  def cancelable[A](k: (Either[Throwable, A] => Unit) => IO[Unit]): F[A]

  def uncancelable[A](fa: F[A]): F[A]

  def onCancelRaiseError[A](fa: F[A], e: Throwable): F[A]

  def start[A](fa: F[A]): F[Fiber[F, A]]

}
```

Details about these operations:

1. `cancelable` is the builder we need, as an alternative to `async`, for specifying cancelation logic (in the returned `IO[Unit]`, that being the cancelation logic suspended with `IO`)
2. `uncancelable` undoes what `cancelable` is doing, thus nullifying any specified cancelation logic, yielding an `F` value that's atomic, i.e. either all of it executes, or none of it
3. `onCancelRaiseError` materialises cancelation with a specified error, also cutting the connection between the producer and the consumer even for atomic (non-cancelable) tasks — this operation is needed for implementing logic like `bracket`
4. `start` will start the evaluation, suspended in the `F` context, yielding a `Fiber` that can be joined or canceled

Note that I've expressed `start` in `Concurrent` as well. I no longer wish to separate them without a clear use-case — like in the `FlatMap` versus `Monad` example, there is a clear use-case for `FlatMap`.

Also, in comparison with the proposal in #126 there is now a much stronger law that requires that the cancellation logic gets triggered on cancellation:

```scala
  def asyncCancelableReceivesCancelSignal[A](a: A, f: (A, A) => A) = {
    val lh = F.suspend {
      var effect = a
      val async = F.cancelable[Unit](_ => IO { effect = f(effect, a) })
      F.start(async).flatMap(_.cancel).map(_ => effect)
    }
    lh <-> F.pure(f(a, a))
  }
```

Our `LTask` implementation in #126 would fail this law. So if you have cancellation logic specified in the `cancelable` builder, then now its usage is required when `cancel` is triggered (and by extension by the token returned via `runCancelable`).

And then the `ConcurrentEffect` type class is obviously:

```scala
trait ConcurrentEffect[F[_]] extends Concurrent[F] with Effect[F] {

  def runCancelable[A](fa: F[A])(cb: Either[Throwable, A] => IO[Unit]): IO[IO[Unit]]
}
```

---

## Missing

N.B. we are missing a `race` operation in this type class hierarchy and we are missing it on `IO` as well. But I think we can add that in a later PR if we agree on this one.

Also `Bracket` is described in #113 and not part of this PR.

---

## The Case for Separating Async from Concurrent

*Given in a comment below, that I'm copy pasting here ...*

The laws do not guarantee that any value is cancelable, but it does guarantee that the cancelation signal is sent and this is enough for the following.

Suppose you want to describe this operation, which is covered by our new [Timer](https://github.com/typelevel/cats-effect/pull/132):

```scala
def sleep(d: FiniteDuration): F[Unit]
```

This would get implemented by [ScheduledExecutorService.schedule](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledExecutorService.html#schedule(java.util.concurrent.Callable,%20long,%20java.util.concurrent.TimeUnit)) or by JavaScript's [setTimeout](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setTimeout). But in both those cases you get a cancelable token that can be used to cancel that scheduled tick in race conditions.

Suppose we do this:

```scala
val timeout = Timer[IO].sleep(10.seconds) *> IO.raiseError(new TimeoutException)

// We don't have race yet, but we will ;-)
IO.race(request, timeout)
```

In my experience, most of the time the happy path will win, otherwise there's something wrong with the system. But in this case if that `F[Unit]` is not cancelable, this creates a resource leak, as the underlying scheduler will end up executing a lot of `Runnable` objects for nothing.

In the past I worked on such a system that was using Scala's `Future` and I did something similar to the above, with the system crashing because it was receiving over 10,000 requests per second, with the scheduler choking on those timeouts. It's also why there's a [Future.timeout](https://github.com/monix/monix/blob/v3.0.0-M3/monix-execution/shared/src/main/scala_2.12/monix/execution/FutureUtils.scala#L38) described in Monix, that is specialized for this specific use-case (instead of using a generic `Future.race`) and what motivated me to work on a cancelable Monix `Task`.

So I think I would be fine with #126, however we need to be clear that the signature above is *unsafe*. The safe version for a non-cancelable `F` is this:

```scala
def sleep(d: FiniteDuration): F[(F[Unit], F[Unit])]
```

So you need one `F[Unit]` for waiting on the delayed tick and another `F[Unit]` for triggering a cancellation. And if you'll notice, this is effectively the shape of our new [Fiber](https://github.com/typelevel/cats-effect/blob/master/core/shared/src/main/scala/cats/effect/Fiber.scala#L50), which is nothing more than this tuple.

Speaking in terms of type signatures, with a `Concurrent` any `F[A]` is actually an `F[Fiber[F, A]]` (`F[(F[A], F[Unit])]`) under the hood, but that is not true with a plain `Async` data type and so when you receive an `F[A]`, its cancelation token is lost and cannot be recovered.

Therefore if we want to provide a safe `sleep` operation, we're talking of two distinct signatures, therefore this logic cannot be implemented with a `Priority`:

```scala
def sleep(d: FiniteDuration)(implicit F: Concurrent[F]): F[Unit]

def sleep(d: FiniteDuration)(implicit F: Async[F]): F[(F[Unit], F[Unit])]
```

Also note that we can find this signature in FS2 actually, in [Promise.cancelableGet](https://github.com/functional-streams-for-scala/fs2/blob/v0.10.2/core/shared/src/main/scala/fs2/async/Promise.scala#L44), a signature that wouldn't be necessary if `F` is cancelable.

---

## Backwards Compatibility Breakage - Analysis

I've ran MiMa against version 0.9 (the currently deployed version), so measuring everything related to the new cancelable IO, including `Timer`, to know what to expect.

I then fixed the incompatibilities. At this point we have this list for version 0.9:

```scala
// Not a problem: AsyncInstances is a private class, we just moved those
// directly in the companion object.
//
// Manually checked:
//  - catsEitherTAsync
//  - catsOptionTAsync
//  - catsStateTAsync
//  - catsWriterTAsync
exclude[MissingClassProblem]("cats.effect.AsyncInstances"),
// Not a problem: AsyncInstances is a private class, WriterTAsync too
exclude[MissingClassProblem]("cats.effect.AsyncInstances$WriterTAsync"),
// Not a problem: AsyncInstances is a private class, OptionTAsync too
exclude[MissingClassProblem]("cats.effect.AsyncInstances$OptionTAsync"),
// Not a problem: AsyncInstances is a private class, EitherTAsync too
exclude[MissingClassProblem]("cats.effect.AsyncInstances$EitherTAsync"),
// Not a problem: AsyncInstances is a private class, StateTAsync too
exclude[MissingClassProblem]("cats.effect.AsyncInstances$StateTAsync"),
//
// Not a problem: EffectInstances is a private class and we just moved
// those in the companion object.
//
// Manual check for:
//  - catsEitherTEffect
//  - catsStateTEffect
//  - catsWriterTEffect
//  - catsWriterTEffect
exclude[MissingClassProblem]("cats.effect.EffectInstances"),
// Not a problem: Missing private traits being inherited
exclude[MissingTypesProblem]("cats.effect.Effect$"),
//
// Not a problem: SyncInstances is a private class, we just moved those
// directly in the companion object.
//
// Manual check for:
//   - catsEitherTEvalSync
//   - catsEitherTSync
//   - catsOptionTSync
//   - catsStateTSync
//   - catsWriterTSync
exclude[MissingClassProblem]("cats.effect.SyncInstances"),
// Not a problem: no longer implementing private traits
exclude[MissingTypesProblem]("cats.effect.Sync$"),
// Not a problem: SyncInstances and StateTSync are private.
exclude[MissingClassProblem]("cats.effect.SyncInstances$StateTSync"),
// Not a problem: SyncInstances and OptionTSync
exclude[MissingClassProblem]("cats.effect.SyncInstances$OptionTSync"),
// Not a problem: SyncInstances and EitherTSync
exclude[MissingClassProblem]("cats.effect.SyncInstances$EitherTSync"),
// Not a problem: SyncInstances and WriterTSync are private
exclude[MissingClassProblem]("cats.effect.SyncInstances$WriterTSync"),
//
// Not a problem: LiftIOInstances is a private class, we just moved
// those directly in the companion object.
//
// Manual check for:
//   - catsEitherTLiftIO
//   - catsKleisliLiftIO
//   - catsOptionTLiftIO
//   - catsStateTLiftIO
//   - catsWriterTLiftIO
exclude[MissingTypesProblem]("cats.effect.LiftIO$"),
exclude[MissingClassProblem]("cats.effect.LiftIOInstances"),
exclude[MissingClassProblem]("cats.effect.LiftIOInstances$OptionTLiftIO"),
exclude[MissingClassProblem]("cats.effect.LiftIOInstances$KleisliLiftIO"),
exclude[MissingClassProblem]("cats.effect.LiftIOInstances$EitherTLiftIO"),
exclude[MissingClassProblem]("cats.effect.LiftIOInstances$StateTLiftIO"),
exclude[MissingClassProblem]("cats.effect.LiftIOInstances$WriterTLiftIO"),
//
// Following are all internal implementation details:
//
// Not a problem: IO.Async is a private class
exclude[IncompatibleMethTypeProblem]("cats.effect.IO#Async.apply"),
// Not a problem: IO.Async is a private class
exclude[MissingTypesProblem]("cats.effect.Async$"),
// Not a problem: IO.Async is a private class
exclude[IncompatibleResultTypeProblem]("cats.effect.IO#Async.k"),
// Not a problem: IO.Async is a private class
exclude[IncompatibleMethTypeProblem]("cats.effect.IO#Async.copy"),
// Not a problem: IO.Async is a private class
exclude[IncompatibleResultTypeProblem]("cats.effect.IO#Async.copy$default$1"),
// Not a problem: IO.Async is a private class
exclude[IncompatibleMethTypeProblem]("cats.effect.IO#Async.this"),
// Not a problem: RestartCallback is a private class
exclude[DirectMissingMethodProblem]("cats.effect.internals.IORunLoop#RestartCallback.this"),
// Not a problem: IOPlatform is private
exclude[DirectMissingMethodProblem]("cats.effect.internals.IOPlatform.onceOnly"),
// Not a problem: IORunLoop is private
exclude[MissingClassProblem]("cats.effect.internals.IORunLoop$RestartCallback$")
```

The generated incompatibilities are generated by getting rid of `*Instances` private classes, so migrating from `object Async extends AsyncInstances` to just `object Async`.

This was unfortunately needed because the compiler crashes due to those instances being expressed in classes, possibly in combination with the SI-2712 fix 😐

I am not sure if this doesn't break binary compatibility. As an educated guess, it shouldn't and it's also the kind of work we did in the `0.8` to `0.9` migration, but MiMa cannot tell us that - we'll need to double check.

SystemFw commented:
The thing I'm not sure about is the separate hierarchy for Cancellable things

alexandru commented:
@SystemFw so you like #126 better? Could you toy with the idea please, in the context of FS2 for example and tell me which one you like best?

SystemFw commented:
First of all, I have to agree with you that this is really hard to get right :)
So I'm more sure about which questions I want to ask, than about answers I'm ready to give. In this case, the question is: what does a non cancellable hierarchy bring to the table, to offset the greater number of concepts?

I'll look at both proposal in more detail, but it's a very busy time unfortunately :(
My first impression is that #126 with `start` and `cancellable` merged into one might be enough, but I have not given it enough thought to be sure.

In the meantime, I think a big thanks is in order, given the incredible amount of work you're doing (particularly the "boring" bits like scaladoc, which looks awesome)

rossabaker commented:
Agree wholeheartedly on the thanks.  @alexandru is doing great work here.

A guarantee that the cancelation action is triggered does not guarantee that the effect is canceled before it runs.  Since it's a best effort, I still don't see why I would ever ask for a `CAsync` over a `Priority[CAsync, Async]`.  And if real world use always demands the latter, it seems much simpler to put it on `Async`.

I guess I'd like to see an example of why I care that it _tried_ to cancel, even if it couldn't, to prefer this over #126.

johnynek commented:
I prefer this design. Bikeshed: i would prefer AsyncWithCancellation or CancellableAsync than just using C.

gvolpe commented:
I do as well want to give you a massive thanks for your incredible amount of work @alexandru ! 

Secondly, I prefer this proposal over #126 given its clear separation of concerns and laws support of the newly introduced typeclasses but I agree with @johnynek that I would prefer the names to be both `CancellableAsync` and `CancellableEffect`, even though it requires a bit more of typing.

alexandru commented:
@johnynek @gvolpe I renamed those classes into `CancelableAsync` and `CancelableEffect` (single "l", because [I dislike](https://twitter.com/alexelcu/status/967704156977934336) the British spelling 😜).

---

@rossabaker I tried giving an example in the ScalaDoc actually, not sure if it's a strong enough example. So the laws do not guarantee that any value is cancelable, but it does guarantee that the cancelation signal is sent and this is enough for the following.

Suppose you want to describe this operation, which is covered by our new [Timer](https://github.com/typelevel/cats-effect/pull/132):

```scala
def sleep(d: FiniteDuration): F[Unit]
```

This would get implemented by [ScheduledExecutorService.schedule](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledExecutorService.html#schedule(java.util.concurrent.Callable,%20long,%20java.util.concurrent.TimeUnit)) or by JavaScript's [setTimeout](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setTimeout). But in both those cases you get a cancelable token that can be used to cancel that scheduled tick in race conditions.

Suppose we do this:

```scala
val timeout = Timer[IO].sleep(10.seconds) *> IO.raiseError(new TimeoutException)

// We don't have race yet, but we will ;-)
IO.race(request, timeout)
```

In my experience, most of the time the happy path will win, otherwise there's something wrong with the system. But in this case if that `F[Unit]` is not cancelable, this creates a resource leak, as the underlying scheduler will end up executing a lot of `Runnable` objects for nothing.

In the past I worked on such a system that was using Scala's `Future` and I did something similar to the above, with the system crashing because it was receiving over 10,000 requests per second, with the scheduler choking on those timeouts. It's also why there's a [Future.timeout](https://github.com/monix/monix/blob/v3.0.0-M3/monix-execution/shared/src/main/scala_2.12/monix/execution/FutureUtils.scala#L38) described in Monix, that is specialized for this specific use-case (instead of using a generic `Future.race`) and what motivated me to work on a cancelable Monix `Task`.

So I think I would be fine with #126, however we need to be clear that the signature above is *unsafe*. The safe version for a non-cancelable `F` is this:

```scala
def sleep(d: FiniteDuration): F[(F[Unit], F[Unit])]
```

So you need one `F[Unit]` for waiting on the delayed tick and another `F[Unit]` for triggering a cancellation. And if you'll notice, this is effectively the shape of our new [Fiber](https://github.com/typelevel/cats-effect/blob/master/core/shared/src/main/scala/cats/effect/Fiber.scala#L50), which is nothing more than this tuple.

Speaking in terms of type signatures, with a `CancelableAsync` any `F[A]` is actually an `F[Fiber[F, A]]` (`F[(F[A], F[Unit])]`) under the hood, but that is not true with a plain `Async` data type and so when you receive an `F[A]`, its cancelation token is lost and cannot be recovered.

Therefore if we want to provide a safe `sleep` operation, we're talking of two distinct signatures, therefore this logic cannot be implemented with a `Priority`:

```scala
def sleep(d: FiniteDuration)(implicit F: CancelableAsync[F]): F[Unit]

def sleep(d: FiniteDuration)(implicit F: Async[F]): F[(F[Unit], F[Unit])]
```

Also note that we can find this signature in FS2 actually, in [Promise.cancelableGet](https://github.com/functional-streams-for-scala/fs2/blob/v0.10.2/core/shared/src/main/scala/fs2/async/Promise.scala#L44), a signature that wouldn't be necessary if `F` is cancelable.

gvolpe commented:
Fair enough, you've saved one word!!! :rofl:

rossabaker commented on core/shared/src/main/scala/cats/effect/CancelableAsync.scala:
```diff
@@ -0,0 +1,327 @@
+import scala.util.Either
+
+/**
+ * Type class for [[Async]] data types that are cancellable.
```
`cancelable`, here and elsewhere.

rossabaker commented on core/shared/src/main/scala/cats/effect/Async.scala:
```diff
@@ -19,118 +19,211 @@ package effect
+ * N.B. this type class in particular is NOT describing cancelable
+ * async processes, see the [[CancelableAsync]] type class for that.
+ *
+ * ==Async Type-class==
```
Nit: "type class" (as elsewhere in the doc) or "typeclass", but probably not "type-class".

rossabaker commented on core/shared/src/main/scala/cats/effect/Async.scala:
```diff
@@ -19,118 +19,211 @@ package effect
-   * Creates an `F[A]` instance from a provided function
-   * that will have a callback injected for signaling the
-   * final result of an asynchronous process.
+   * Creates a simple, non-cancelable `F[A]` instance that
```
If we're being American about this, [noncancelable](https://www.merriam-webster.com/dictionary/noncancelable) is unhyphenated.  (And "uncancelable" isn't a word, so maybe that was a better name for the method on `IO`.)

rossabaker commented on core/shared/src/main/scala/cats/effect/Async.scala:
```diff
@@ -19,118 +19,211 @@ package effect
+   *
+   * N.B. expressing this conversion in terms of `Async` and its
+   * capabilities means that the resulting `F` is not cancelable.
+   * [[CancelableAsync]] then overloads this with an implementation that is.
```
s/overloads/overrides/

rossabaker commented on core/shared/src/main/scala/cats/effect/CancelableAsync.scala:
```diff
@@ -0,0 +1,327 @@
+ * Java's `ScheduledExecutorService#schedule`, which will return a
+ * Java `ScheduledFuture` that has a `.cancel()` operation on it.
+ *
+ * Similarly, for `CAsync` data types, we can provide cancellation
```
`CancelableAsync`, here and elsewhere

rossabaker commented on core/shared/src/main/scala/cats/effect/CancelableEffect.scala:
```diff
@@ -0,0 +1,118 @@
+ * also returns a token that can be used for cancelling the running
+ * computation.
+ *
+ * Note this is the safe and generic version of [[IO.unsafeRunCancelable]].
```
I don't understand this comment.

alexandru replied:
I wanted to say that you can safely do ... `runCancelable.unsafeRunSync.unsafeRunSync`, without blocking any threads, because just as with `runAsync` the returned `IO` only starts execution and then it only sends the cancel signal, but without any blocking for some result.

rossabaker commented on core/shared/src/main/scala/cats/effect/IO.scala:
```diff
@@ -319,34 +378,10 @@ sealed abstract class IO[+A] {
   /**
    * Converts the source `IO` into any `F` type that implements
-   * the [[cats.effect.Async Async]] type class.
+   * the [[cats.effect.CancelableAsync CAsync]] type class.
```
To the `LiftIO` typeclass?

alexandru replied:
Fixed.

rossabaker commented on core/shared/src/main/scala/cats/effect/IO.scala:
```diff
@@ -404,7 +439,7 @@ private[effect] abstract class IOInstances extends IOLowPriorityInstances {
   }
 
-  implicit val ioEffect: Effect[IO] = new Effect[IO] {
+  implicit val ioEffect: CancelableEffect[IO] = new CancelableEffect[IO] {
```
Is it worth renaming this to `ioCancelableEffect`?

alexandru replied:
I don't think so. Another way of looking at it is that these are the `cats.effect` instances for `IO` 😀

SystemFw commented:
Apologies for the non conceptual point, but I think these little things end up mattering in the every day: polymorphic code in Scala is already pretty verbose (since you can't have synonyms for a bunch of constraints), and `CancellableEffect` is really long, especially considering that it would be the standard in a lot of code .

Since I think the majority of real implementations will be cancellable, I'd say they should have the shortest "default" names, not the other way around, e.g. I'd rather have `Effect` and `NonCancellableEffect` than `CancellableEffect` and `Effect`. This doesn't work out of the box with `Async`, so we'd have to think of something for `CancellableAsync` (`MonadFork`?) but I think it's worth considering.

________________

Apart from that, the comment by @alexandru above gives a good reason to split the cancellable and non cancellable hierarchy when you have NonCancellable data types, but I'm still not sure it's a good enough reason.

In particular, what types would have instances for the non cancellable hierarchy?
And "the current cats-effect IO" is not a great answer for me, since this whole thing started to address non-cancellation of IO, as a specific shortcoming.
I believe the only valid answers will be things like `ConnectionIO`, i.e. things that are `Async`, but not `Effect`, and therefore that you will never wish for an `Effect` without `start` and `cancellation`, which is a consequence of the fact that you can express both the operations defined in CancellableAsync, _in a worse way_, using the current `Effect` only, which is what we do in fs2 now.

pchlupacek commented:
I fully agree with @SystemFw observation. In case of fs2 I think most of the places where we use Effect will be now subject of CxxEffect, and I think this is really confusing. I agree that cancellation shall be default, and `NonC` variants shall be explicitly named.

gvolpe commented:
I think @SystemFw has a good point. If breaking changes are being introduced, then making an `Effect` cancellable by default makes sense and non-cancellable effects should be explicitly named (eg. `NonCancellableEffect`).

Last but not least, I think the American naming is not consistent with naming in the Cats library, For example we already have `cats.Parallel`, `cats.NonEmptyParallel` and probably others I can't think of right now.

johnynek commented:
I’m confused why Parallel is inconsistent with American spelling. Is there another way to spell parallel that is more common in the US? I’m not aware of such.

alexandru commented:
> Since I think the majority of real implementations will be cancellable, I'd say they should have the shortest "default" names, not the other way around, e.g. I'd rather have Effect and NonCancellableEffect than CancellableEffect and Effect. This doesn't work out of the box with Async, so we'd have to think of something for CancellableAsync (MonadFork?) but I think it's worth considering.

I'm 👎 on a rename. For one, such a rename is going to cause more breakage in the ecosystem than necessary.

But also, I think `Async` will end up being used more in polymorphic code than `CancelableAsync`.
When it comes to parametric polymorphism restricted via type classes, the principle of least power is applied - you only request what you need. And not all async operations that you want described are cancelable.

> In particular, what types would have instances for the non cancellable hierarchy? And "the current cats-effect IO" is not a great answer for me, since this whole thing started to address non-cancellation of IO, as a specific shortcoming.

I think the current `cats.effect.IO` is proof enough that such a type can exist and was and might still be desirable. [Daniel makes good points](https://github.com/typelevel/cats-effect/issues/93#issuecomment-349720286) about it.

Not only that, but `monix.tail.Iterant` can implement `Async`, but not `CancelableAsync`. See https://github.com/monix/monix/pull/600 - not sure if this is any useful for users, beats me, but for the fact that it can, that's a good enough reason to do it.

Btw, I must repeat the obvious — the `Effect` type class has been exactly as powerful as the `IO` (up until 0.9) and the `Async` type class describes types that are at least as powerful as `IO`. This has been bad for Monix's `Task` because at the moment of writing, polymorphic code cannot use `Task`'s capabilities. Also see [this discussion on ToIO](https://github.com/typelevel/cats-effect/issues/73#issuecomment-322056287)

Lets not restrict the type class hierarchy to types that are exactly as powerful as `IO` because then it becomes a useless type class hierarchy.

SystemFw commented:
> I'm 👎 on a rename. For one, such a rename is going to cause more breakage in the ecosystem than necessary.

How so? I'm not proposing to rename any of the existing typeclasses, I'm proposing to rename the new ones, and add operations to the existing ones. In fact, it's the opposite: fs2 and http4s will break everywhere, having to use `CancellableEffect` now. 

> But also, I think Async will end up being used more in polymorphic code than CancelableAsync and Effect will end up being used far more than CancelableEffect - a guesstimate for sure, but it's an educated one

I don't agree with this. Ime either you need `Sync` only, or `Effect` when concurrency is involved, which will now have to be `CancellableEffect`. Again, http4s and fs2 are two concrete cases, so mine is more than a guess.

> I think the current cats.effect.IO is proof enough that such a type can exist and was and might still be desirable. Daniel makes good points about it. Therefore a "LightIO" might still happen.

Fair enough. I don't think it will given the reasons I've already given elsewhere (that convinced you to do this after you stumbled in the same issues, with Scheduler :P)

> There's also Scala's Future. Not exactly the lawful type you want to hear about, but it's an example of an abstraction that describes asynchronous processes as well and is not cancelable, i

I fail to see how this is relevant at all, given that Future cannot even be a Sync (thankfully).

> Also monix.tail.Iterant can and now does implement Async, but not CancelableAsync.

But that's what I'm saying (I've used ConnectionIO as another example), I think that the split is not Cancellable or not, it's Async vs everything else down below. And the reason for this is that if you can have `Effect`, you can also have `fork` and concurrency (as we do now), just in a worse way, so I don't think you'll have an Effect that isn't best served by also not being `CancellableAsync` and `CancellableEffect`. Iterant is not a counter example, since it can't be Effect.

alexandru commented:
@SystemFw 

Well, another reason is that a new name like `NoncancelableAsync` is incorrect. Because this `Async` is not describing `F[_]` data types that are noncancelable, it just describes the `async` operation for building noncancelable `F[_]` values, which is reflected in the name of the type class.

And then `CancelableAsync` has the `cancelable` builder. If we want to keep naming consistency, then these operations will have to be renamed too.

> In fact, it's the opposite: fs2 and http4s will break everywhere, having to use CancellableEffect now.

Which is absolutely normal, because you're changing the semantics of your operations. When we break people's code, it's best if the compiler signals it via new names. Breaking compatibility by changing existing names is a big no-no.

I'm also sympathetic to Rich Hickey's viewpoint that compatibility should never be broken without changing the namespace (in our case `cats.effect` -> `cats.effect2` or something 😉), a viewpoint he explains in [Spec-ulation](https://www.youtube.com/watch?v=oyLBGkS5ICk).

> I fail to see how this is relevant at all, given that Future cannot even be a Sync (thankfully).

It's very relevant for its shape. Not all asynchronous operations can be described with `(A => Unit) => Cancelable`, some of them are `(A => Unit) => Unit` and this is the more generic signature.

> I don't agree with this. Ime either you need Sync only, or Effect when concurrency is involved ...

As a counter example, absolutely all operations that wrap the blocking I/O operations in `java.io` will be `Async` only, anything that involves simple thread shifting actually, introduced for fairness reasons, because they execute as atomic operations:

```scala
def readLine[F[_]](in: BufferedReader)
  (implicit F: Async[F], ec: ExecutionContext): F[String] =
  Async.shift *> F.delay(blocking(in.readLine()))
```

SystemFw commented:
> Well, another reason is that a new name like NoncancelableAsync is incorrect.

Which is why I said:

> This doesn't work out of the box with Async, so we'd have to think of something for CancellableAsync

> It's very relevant for its shape. Not all asynchronous operations can be described with (A => Unit) => Cancelable, some of them are (A => Unit) => Unit

I think this is arguing against something I've never proposed :)

To be clear, what I'm provisionally proposing is this:

- Async stays exactly as it is
- CancellableAsync is renamed to something shorter (`MonadFork`?).  This will be super common in ops that support concurrency, and typeclass constraints propagate the stronger one: things at the bottom of the stack like fs2 will need this, and everything else will likely have to use them as well when calling fs2 code. This phenomenon happens today with Effect.
- CancellableEffect is just Effect.
- NonCancellableEffect will see no use in my opinion, but I don't mind keeping it to support a LightIO type like today's IO. It shouldn't have the default name though, it will have zero instances when we release, for one.

> Which is absolutely normal, because you're changing the semantics of your operations. When we break people's code, it's best if the compiler signals it via new names. Breaking compatibility by changing existing names is a big no-no.

I don't see how this specific case is any different from every other case in the Scala ecosystem where we didn't apply this principle. In particular, we did not rename IO to CancellableIO when the cancellation PR went in: here we use versioning to signal breaking changes. 

> As a counter example, absolutely all operations that wrap the blocking I/O operations in java.io will be Async only

Sure, which is why I'm not proposing to touch Async at all.

alexandru commented:
Also, let me give an `Effect` sample that will remain `Effect`, the [Observable#mapEval](https://github.com/monix/monix/blob/master/monix-reactive/shared/src/main/scala/monix/reactive/Observable.scala#L1534):

```scala
abstract class Observable[+A] {

  def mapEval[F[_], B](f: A => F[B])(implicit F: Effect[F]): Observable[B]
}
```

This operation cannot be described with just `Sync`, like in [Iterant#mapEval](https://github.com/monix/monix/blob/v3.0.0-M3/monix-tail/shared/src/main/scala/monix/tail/Iterant.scala#L761), because it is not `F[_]` that drives the evaluation, but `Observable` itself.

And we cannot use `CancelableEffect` here, due to Observable's underlying protocol, which is described by [Observer](https://github.com/monix/monix/blob/v3.0.0-M3/monix-reactive/shared/src/main/scala/monix/reactive/Observer.scala#L47):

```scala
trait Observer[-A] extends Any with Serializable {

  def onNext(elem: A): Future[Ack]

  def onError(ex: Throwable): Unit
  def onComplete(): Unit
}
```

So because we're using `Future[Ack]` in `onNext`, we have no need for a cancelable `Effect`, because `Future` is not cancelable. And note this has been by choice, because we also have a memory safe and sweat [monix.execution.CancelableFuture](https://github.com/monix/monix/blob/v3.0.0-M3/monix-execution/shared/src/main/scala/monix/execution/CancelableFuture.scala) implementation.

In this protocol `Future` is used for back-pressure and only for back-pressure. Having a `CancelableFuture` in there would be at the same time next to useless and also insane. Having to deal with a cancelable future in this protocol is insane, due to its (hybrid) push-based nature.

Also note that `Observable` could implement `CancelableAsync`, if it had a memory safe `flatMap`. It already has a memory safe `++`, so a memory safe `flatMap` might still happen. So this has nothing to do with its cancelability capabilities, rather it has to do with not outsourcing such capabilities.

SystemFw commented:
Could you point me to what type could have an instance of  a non cancellable Effect when we will release?

alexandru commented:
On the proposal:

> CancellableAsync is renamed to something shorter 

Which is why it was originally named `CAsync`. If you want, we can revert to that.

I do not want `MonadFork`, because this type class does not describe `start`. `start` is there as a bonus. I'm actually pretty against usage of `start`, Monix's `Task` being usable just fine without it, `start` being now there due to popular demand.

> CancellableEffect is just Effect

No, because that changes the semantics, it changes the meaning of the name and that's now how type-class based design is performed. Cancelability is a capability built on top of non-cancelable abstractions, not the other way around.

> NonCancellableEffect will see no use in my opinion

I just gave you a perfectly valid example above. I can probably come up with more.

> I don't see how this specific case is any different from every other case in the Scala ecosystem where we didn't apply this principle. In particular, we did not rename IO to CancellableIO when the cancellation PR went in: here we use versioning to signal breaking changes.

This is because the `IO` simply provides more capabilities without changing the semantics of the existing operations.  The new `IO` is at least source compatible with the old `IO` (the PR introducing it in master was actually binary compatible, this PR breaking that due to its `#toIO` method, which is a good thing because it makes it more generic). And you can see this in its design — for example in the fact that cancellation is always opt-in.

Adding capabilities is *accretion*, changing the semantics of an existing name is *breakage*.
We're talking accretion vs breakage.

SystemFw commented:
> I just gave you a perfectly valid example above. I can probably come up with more.

Well, I didn't really understand the example, but I'm sure it is valid if you say so, so I'm ready to concede that "will see no use" is wrong.
 Still, I don't think it should be the _default_, cause there are many more examples that will have to change. I'm also still waiting for an example of a type that will have (when we release) an instance of Effect but not CancellableEffect.

> I do not want MonadFork, because this type class does not describe start. start is there as a bonus. 

Well, that's a broader disagreement there then. We _need_ something like `fork: F[A] => F[Unit]` to implement concurrency, we are now implementing it with `runAsync` which is why `Effect` is so pervasive.  

> No, because that changes the semantics, it changes the meaning of the name and that's now how type-class based design is performed.

I wasn't aware you were never allowed to break things when using typeclasses. I agree it's a big break, but so is the type class redesign as a whole, and this is our chance to get things right before 1.0

SystemFw commented:
In any case, let's see what the others think as well, @djspiewak and @mpilquist in particular haven't weighed in at all, I'm looking forward to hearing what they have to say :)

alexandru commented:
> Well, that's a broader disagreement there then. We need something like fork: F[A] => F[Unit] to implement concurrency, we are now implementing it with runAsync which is why Effect is so pervasive.

I have expressed my dislike of `start` here: https://github.com/typelevel/cats-effect/issues/88#issuecomment-355947347

Basically `CancelableAsync` is about the `cancelable` builder. The `start` operation is there because many people think it's useful as well and in support of the `cancelable` builder - because you can't describe laws for that `cancelable` builder without having a way to trigger cancellation.

I'm not saying that you should not use `start` in FS2 or whatever. I'm saying that `CancelableAsync` is not about `start`.

> I wasn't aware you were never allowed to break things when using typeclasses. I agree it's a big break, but so is the type class redesign as a whole, and this is our chance to get things right before 1.0

We haven't released 1.0, but the `cats-effect` project is already used in production code and we can't hide behind the umbrella of the 0.x releases when breaking things, because it will cause suffering nonetheless. We can break things, the new `IO` is definitely a different beast than the old `IO`, but when we do, we need a really good reason to do it. I fear for example that I broke too many things in the upcoming Monix 3.0 release, but I think I can justify those changes with the migration to `cats-effect`.

And at this point I think this design is the correct one. Too many type classes certainly impact usability, but then requiring *just the needed capabilities* in code making use of constrained parametric polymorphism is the advantage of using type classes. So I can see users requiring `CancelableAsync` or `CancelableEffect` only when they need to deal with cancellation.

I dislike the long names for sure, which is why originally I had `CAsync` and `CEffect`, so if you can think of other names, now would be the time.

> In any case, let's see what the others think as well, @djspiewak and @mpilquist in particular haven't weighed in at all, I'm looking forward to hearing what they have to say :)

Me too. I do have a feeling that @djspiewak disagrees with this whole migration towards a cancelable `IO`. He's either very busy or he lost hope in this project. I hope it's the former and not the later 😐

SystemFw commented:
Just a couple quick points, since I have a plane to catch and we have probably exposed the respective arguments well enough for others to form an opinion.

> So I can see users requiring CancelableAsync or CancelableEffect only when they need to deal with cancellation.

This is not what's going to happen. We will need CancellableAsync for `start` (cancellation or not) just like we need Effect for `fork` now, and we are at the bottom of the stack, so users will have to use that constraint regardless.

> Me too. I do have a feeling that @djspiewak disagrees with this whole migration towards a cancelable IO. He's either very busy or he lost hope in this project.

I share the same feeling :( Let's hope for the best.

alexandru commented:
> This is not what's going to happen. We will need CancellableAsync for start (cancellation or not) just like we need Effect for fork now, and we are at the bottom of the stack, so users will have to use that constraint regardless.

If you'll want `start`, then yes, you need `CancellableAsync`. I don't want us to describe `start` for `F[_]` data types that are not cancelable, because that's unsafe. This is why, once `CAsync` was in, it made no sense to provide a separate `AsyncStart` as in the other PR ... until people will express their desire for it.

But note that not all concurrent or async operations require `start`, only those where you want to describe race conditions. We could build a perfectly safe `parMap2` for the old `IO` for example.

SystemFw commented:
Oh, I very much agree that start and cancellation are linked, which is why the current implementation based on effect is not satisfactory.
But then, just as you can say that the type class is really about cancel, with start thrown in there, I can say that is really about start, with cancel being the unavoidable corollary. 

And under this perspective, it would be much more pervasive. It's true that you can implement parMap2 without it, but that's because that's really parallelism via concurrency, actual concurrency requires it, in a polymorphic context

alexandru commented:
>  But then, just as you can say that the type class is really about cancel, with start thrown in there, I can say that is really about start, with cancel being the unavoidable corollary.

Interestingly, you're right. This is how in math you have to pick your axioms and then you can sometimes switch some theorem for an axiom, with the theory built on top of that still making sense 🙂

SystemFw commented:
@alexandru Proposed `Concurrent` instead of `CancellableAsync` on Gitter. I love that.

alexandru commented:
UPDATE:

I have renamed `CancelableAsync` to `Concurrent` and `CancelableEffect` to `ConcurrentEffect`, which is still a mouthful, but describes the type very precisely:

```scala
trait ConcurrentEffect[F[_]] extends Concurrent[F] with Effect[F] {

  def runCancelable[A](fa: F[A])(cb: Either[Throwable, A] => IO[Unit]): IO[IO[Unit]]
}
```

Also, a law related to `start` is failing for the `Concurrent[StateT[F, S, ?]]` instance. I'm currently investigating, but it's something really, really weird.

alexandru commented:
UPDATE:

- tests are fixed
- `race` is missing from this hierarchy, but is already prepared to go in a new PR, since this one is already too big (`race` needing an `IO` implementation as well)

Assuming no further fixes or renaming happens, the PR is ready to be merged.

mpilquist commented:
Looks great! :+1:

SystemFw commented:
I don't have the repo rights to approve it, so I'm just giving you an :+1: and another big thanks for all the work :)

Mostly unrelated question, does `race` need `ConcurrentEffect`, or is `Concurrent` enough you reckon?

alexandru commented:
`Concurrent` is enough for `race` and I'm thinking of putting it in `Concurrent` as well ... if more granularity is needed, we can always split `Concurrent` later.

Btw, it's a bad sign when an operation needs `Effect` / `EffectConcurrent`, because `Effect*` is effectively equivalent with `IO`.

So thankfully that's not the case for `race`.

alexandru commented:
In the description of this PR I have included a "*Backwards Compatibility Breakage - Analysis*".

The really cool thing about the changes we're preparing for 0.10 is that the API is still source compatible and if we tried really hard we could even maintain binary compatibility!

But this is a good time to do some cleanup, so some obvious changes were made. See the details in the description.

johnynek commented on core/shared/src/main/scala/cats/effect/ConcurrentEffect.scala:
```diff
@@ -0,0 +1,119 @@
+   *
+   *   // For triggering asynchronous execution
+   *   val cancel = io.unsafeRunSync
+   *   // For cancellation
```
What about an example without unsafeRunSync? For instance, an application that is a single `IO[Unit]` value that is triggered run in a `main` method. It might be nicer to show an example where this is run or canceled without calling unsafe.

LukaJCB commented on core/shared/src/main/scala/cats/effect/IO.scala:
```diff
@@ -404,7 +439,7 @@ private[effect] abstract class IOInstances extends IOLowPriorityInstances {
   }
 
-  implicit val ioEffect: Effect[IO] = new Effect[IO] {
+  implicit val ioEffect: ConcurrentEffect[IO] = new ConcurrentEffect[IO] {
```
Tiny nitpick, maybe rename to `ioConcurrentEffect`?

alexandru replied:
TBH I would rather not. That name can be considered to come from `cats.effect`.

Renaming it breaks binary compatibility. Not such a problem right now, but we need a stable name and we can't keep changing it depending on what type classes `IO` implements.

So it's either `ioEffect` (from `cats.effect`) or something else. Naming is hard, but I'd rather not have this `ioConcurrentEffect`.

LukaJCB replied:
Ah yes, that's a good reason, thanks! 👍

LukaJCB commented on laws/shared/src/main/scala/cats/effect/laws/discipline/Arbitrary.scala:
```diff
@@ -90,6 +90,11 @@ object arbitrary {
 
-  implicit def catsEffectLawsCogenForIO[A](implicit cgfa: Cogen[Future[A]]): Cogen[IO[A]] =
-    cgfa.contramap((ioa: IO[A]) => ioa.unsafeToFuture())
+  implicit def catsEffectLawsCogenForIO[A, B](implicit cga: Cogen[A]): Cogen[IO[A]] =
```
Big 👍

LukaJCB commented on core/shared/src/main/scala/cats/effect/ConcurrentEffect.scala:
```diff
@@ -0,0 +1,119 @@
+   *   cancel.unsafeRunSync
+   * }}}
+   */
+  def runCancelable[A](fa: F[A])(cb: Either[Throwable, A] => IO[Unit]): IO[IO[Unit]]
```
I like the nested `IO`, but it doesn't exactly signal that the inner type means cancelation. This is a pretty low level API so I think it's perfectly fine, but if someone e.g. calls `flatten` on it, it will immediately run and cancel the execution, right? I sort of prefer the `CancelableFuture` that Monix provides, but I still think this API is good enough for what we set out to do. But maybe someone has a good idea that nobody's thought about yet, so I just wanted to provide some food for thought. :)

alexandru replied:
I like Monix's approach too, but there we have the luxury of an entire sub-project supporting it (monix-execution), whereas cats-effect is more minimal.

I think `IO[Unit]` is reasonable though. Once you get accustomed to what cancellation looks like, there's not much that an `IO[IO[Unit]]` can mean.

LukaJCB replied:
Yeah, agreed :)

alexandru commented:
Thanks a lot for the review guys, going to torment you with another commit unfortunately: https://github.com/typelevel/cats-effect/pull/134/commits/076c40985df8cc2f2242565015e1c8faccaad835

I've added to `Concurrent` the missing pieces for being able to express `bracket`:

- `uncancelable`
- `onCancelRaiseError`

The reasoning on `uncancelable` is that once you have a mechanism for specifying cancelation logic, you also need a mechanism for nullifying it. The idea is that once you have an `IO`, sometimes you really want that `IO` to be an *atomic operation*, so either all of it executes, or none of it.

And `onCancelRaiseError` is also an extension needed for expressing `bracket` basically. This is a mechanism for materialising cancellation. In other implementations, like Java's `CompletableFuture` or Haskell's async exceptions, tasks end up being completed in error when canceled. The mechanism I've chosen actually splits this part in two and it's a little different. Lighter and probably more flexible. But this second part is needed for being able to actually detect cancelation, otherwise we can't implement something like `bracket`.

Both these operations were already added in #121 on `IO`.

The good news is that by adding laws in `Concurrent`, I've spotted a problem in the implementation of `IO#start`, when used in combination with `onCancelRaiseError`, which is now fixed.

---

If I'm getting thumbs up from you again, I'd like to have this merged tomorrow night, which would be ideal, as I've got some more free time this week and would like to focus on describing `race` too.

LukaJCB commented:
Haha, reading your comment is never a torment! <3

I'm not quite sure if I understood your explanation for `onCancelRaiseError` being needed for `bracket`, I understand the `uncancelable` part, but maybe I'll get a better grasp of it once I dive into the code!

alexandru commented:
In order to implement something like `bracket`, you need to materialise cancelation somehow. And `onCancelRaiseError` makes this possible, e.g. gave this sample in the ScalaDoc:

```scala
import scala.concurrent.CancellationException
// Reusable instance
val wasCanceled = new CancellationException()

F.onCancelRaiseError(fa, wasCanceled).attempt.flatMap {
  case Right(a) =>
    F.delay(println(s"Success: \$a"))
  case Left(`wasCanceled`) =>
    F.delay(println("Was canceled!"))
  case Left(error) =>
    F.delay(println(s"Terminated in error: \$error"))
}
```

You can also make that canceled branch to be non-terminating after you deal with it 😉

SystemFw commented:
ah, I see you what you mean. 
In haskell to cancel you throw an async exception to a lightweight thread directly.
Here you put the `F` in a cancelled state, and the `F` has an operation to materialise cancellation ("throwing to itself", if you will)

alexandru commented:
@rossabaker I have addressed your binary compatibility concerns and updated the description above.

At this point we only have issues related to the private `Instances` classes, which I hope don't break binary compatibility — those changes were needed, as described, because those mixins were crashing the compiler 😔 plus they aren't needed anyway.

We'll have to double check, but at this point I'm assuming that the 0.10 release is going to be binary compatible (but not source compatible) with version 0.9.

I did so, as you suggested, by introducing an [IOBinaryCompat](https://github.com/alexandru/cats-effect/blob/17c9d338de3f8e2848c5867b08b2f334f5a989bf/core/shared/src/main/scala/cats/effect/internals/IOBinaryCompat.scala#L31) that makes those definitions `private[internals]`, such that the compiler doesn't get confused not even in Cats's own code base.

Same for `Async#shift`, it's now a `private[effect]` with a `@deprecated` annotation too.

alexandru commented:
I have updated the description with the MiMa filter list included now in `build.sbt` to prove compatibility against 0.9.

alexandru commented:
I think concerns have been addressed.

I hope nobody minds that I merge this, as I want to unblock further work on `race`, plus I want to release a hash version in order to test that this release is indeed binary compatible with 0.9.
Until we release 0.10 everything is revertible anyway.

Plus if I kept it open, I fear more commits will come and that's a bad idea for visibility 🙂
